### PR TITLE
gh-130806: emit ResourceWarning if GzipFile unclosed

### PR DESCRIPTION
**Pull request by cmaloney**
This may indicate accidental data loss, the `crc` and `size` fields of the GzipFile are only written on close.

Ways to make sure all data is written:
1. Use the [file-like object](https://docs.python.org/3/glossary.html#term-file-object) as a [“With Statement Context Manager”](https://docs.python.org/3/reference/datamodel.html#context-managers).
   - All objects which [inherit](https://docs.python.org/3/tutorial/classes.html#inheritance) from [IOBase](https://docs.python.org/3/library/io.html#io.IOBase) support this.
   - [`BufferedIOBase`](https://docs.python.org/3/library/io.html#io.BufferedIOBase), [`BufferedWriter`](https://docs.python.org/3/library/io.html#io.BufferedWriter), and [`GzipFile`](https://docs.python.org/3/library/gzip.html#gzip.GzipFile) all support this.
   - Ensures `.close()` is called in both exception and regular cases.
1. Ensure [`.close()`](https://docs.python.org/3/library/io.html#io.IOBase.close) is always called which flushes data before closing.
1. If the underlying stream needs to be kept open, use [`.detach()`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.detach)

Since at least 3.12 flushing has been necessary in GzipFile (see gh-105808 which was a release blocker), this makes that more visible. Users have been encountering as they upgrade to 3.12 (ex. gh-129726).

* Issue: gh-130806

### CODE REVIEW

File: Lib/gzip.py [hunk 1]
```diff
@@ -435,6 +440,13 @@ def readline(self, size=-1):
         self._check_not_closed()
         return self._buffer.readline(size)
 
+    def __del__(self):
+        if self.mode == WRITE and not self.closed:
+            import warnings
+            warnings.warn("unclosed GzipFile",
+                          ResourceWarning, source=self, stacklevel=2)
+
+        return super().__del__()
```
Comment (vstinner) [inline Lib/gzip.py]: I don't think that `__del__()` has a return value.

```suggestion
        super().__del__()
```

File: Lib/gzip.py [hunk 2]
```diff
@@ -358,7 +363,7 @@ def closed(self):
 
     def close(self):
         fileobj = self.fileobj
-        if fileobj is None or self._buffer.closed:
+        if fileobj is None or self._buffer is None or self._buffer.closed:
             return
```
Comment (vstinner) [inline Lib/gzip.py]:
```suggestion
        if fileobj is None:
            return
        if self._buffer is None or self._buffer.closed:
            return
```

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Merged, thank you.
